### PR TITLE
Add docs for RPC, dist autograd, and RRef modules

### LINKED ISSUE #28983: Add document page for torch.distributed.rpc/autograd
`distributed_c10d` has a nice doc page [here](https://pytorch.org/docs/stable/distributed.html), and the source code of that live in [here](https://github.com/pytorch/pytorch/blob/master/docs/source/distributed.rst). We should add a similar page for rpc and distributed autograd to describe the API.

@rohan-varma brought up a point that whether we want to add the new docs into the existing `distributed.rst` file as `distributed/distributed_c10d.py`, `distributed/rpc/` and `distributed/autograd` all live in the same `distributed/` folder, or whether we want to keep them in separate files. I would vote for keeping them separate from the existing one, as the existing one mostly focus on c10d, and the new ones are on rpc/autograd, which are quite different topics. The following items would need more discussion and let's track them here:

1. The title of the existing `distributed.rst` page says **Distributed communication package - torch.distributed**, do we want to rename that to _Distributed collective communication package - torch.distributed_ to be more precise?
2. Should rpc and distributed autograd api docs live in the same file? 

 

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @rohan-varma @xush6528

### PR DESCRIPTION
**Pull request by rohan-varma**
Closes https://github.com/pytorch/pytorch/issues/28983. Documentation for `torch.distributed.rpc` and `torch.distributed.autograd` modules. Also fixes/tidies up some of the docstrings in rpc/autograd, and moves some functions to be private so they don't show up in the documentation.

Note: Much of the text to describe/explain the RPC/RRef layers are taken from the following RFCs: https://github.com/pytorch/pytorch/issues/23110, https://github.com/pytorch/pytorch/issues/26759

### CODE REVIEW

File: docs/source/model_parallel.rst [hunk 1]
```diff
@@ -0,0 +1,11 @@
+.. role:: hidden
+    :class: hidden-section
+
+Distributed RPC Framework - torch.distributed.autograd and torch.distributed.rpc
+=====================================================
+
+.. automodule:: torch.distributed.rpc
+.. currentmodule:: torch.distributed.rpc
+
+.. autofunction:: rpc_sync
```
Comment (pritamdamania87) [inline docs/source/model_parallel.rst]: Is there a way to automatically pull in all public methods of a module instead of specifying each one like this?

File: docs/source/model_parallel.rst [hunk 2]
```diff
@@ -0,0 +1,11 @@
+.. role:: hidden
```
Comment (pritamdamania87) [inline docs/source/model_parallel.rst]: Is this hidden on purpose? I think it should be fine to not hide this for now since it'll only be in master and won't show up in some stable docs.

File: docs/source/model_parallel.rst [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    :class: hidden-section
+
+Distributed RPC Framework - torch.distributed.autograd and torch.distributed.rpc
```
Comment (pritamdamania87) [inline docs/source/model_parallel.rst]: Only `Distributed RPC Framework` should be sufficient, otherwise the entire title is too long for the title.

File: docs/source/model_parallel.rst (hunk 1 above)
Comment (rohan-varma) [reply docs/source/model_parallel.rst]: @pritamdamania87 it looks like other doc-related commits list out each public method, see for example https://github.com/pytorch/pytorch/commit/8915e2710c57cefd4bb5a7876216f08d3bc8c92f#diff-7aa5f4b7f9dedcb7a6ce129b9d3c5eb8. The sphix documentation (http://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html) also makes it seems like this needs to be done - there is support for recursively including members of a class but many of our functions do not belong to a specific class.

File: docs/source/model_parallel.rst (hunk 2 above)
Comment (rohan-varma) [reply docs/source/model_parallel.rst]: Should we keep them hidden if the docs aren't polished completely?

File: docs/source/model_parallel.rst [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+Distributed RPC Framework
+=====================================================
+
+The distributed RPC framework provides mechanisms for multi-machine model training through a set of primitives to allow for remote communication, and a higher-level API to automatically differentiate models split across several machines.
+
+RPC and RRef Framework
+====================================================
+
+Bbasics
```
Comment (mrshenli) [inline docs/source/model_parallel.rst]: Basics

File: torch/distributed/rpc/api.py [hunk 5]
```diff
@@ -223,9 +223,10 @@ def rpc_sync(to, func, args=None, kwargs=None):
                        invocation.
 
     Returns:
-        Returns the result of running ``func``on ``args`` and ``kwargs``.
+        Returns the result of running ``func`` on ``args`` and ``kwargs``.
 
     Example::
+
```
Comment (mrshenli) [inline torch/distributed/rpc/api.py]: Curious: why do we need the new line here?

**Review by mrshenli — COMMENTED** [review]
Thanks for adding this!

Hey @jlin27, should we keep this hidden for now as we haven't finished polishing the docstrings yet?

File: torch/distributed/rpc/api.py (hunk 5 above)
Comment (rohan-varma) [reply torch/distributed/rpc/api.py]: Adding this makes the code appear nicely and in python format on the documentation page

File: docs/source/model_parallel.rst [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+RPC and RRef Framework
+====================================================
+
+Basics
+------
+
+Before using RPC and distributed autograd primitives, initialization must take place. First, a backend over which RPCs can be sent over must be initialized. The default (and currently, only available) implementation is the `ProcessGroup` backend, and must be initialized with `torch.distributed.init_process_group` before using other function. See the `documentation for torch.distributed <https://pytorch.org/docs/stable/distributed.html>`_ for additional details. Next, the local RPC agent can be initialized, after which the process will be able to send and receive RPCs from all other connected processes.
+
+.. automodule:: torch.distributed.rpc
+    :members:
```
Comment (rohan-varma) [inline docs/source/model_parallel.rst]: @pritamdamania87 this will add all members of this module, but we'd want to be careful in using this since non-polished/private functions could show up.

File: docs/source/model_parallel.rst (hunk 2 above)
Comment (jlin27) [reply docs/source/model_parallel.rst]: (Also in response to similar question from @mrshenli) 
As @pritamdamania mentioned, it's fine to have unfinished docs out in master; however, tagging them as hidden can be a helpful way to keep track of what still needs work prior to the final release.

File: docs/source/rpc.rst [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+Before using RPC and distributed autograd primitives, initialization must take place. First, a backend over which RPCs can be sent over must be initialized. The default (and currently, only available) implementation is the `ProcessGroup` backend, and must be initialized with `torch.distributed.init_process_group` before using other function. See the `documentation for torch.distributed <https://pytorch.org/docs/stable/distributed.html>`_ for additional details. Next, the local RPC agent can be initialized, after which the process will be able to send and receive RPCs from all other connected processes.
+
+.. automodule:: torch.distributed.rpc
+    :members:
+
+RRef
+------
+
+An RRef (Remote REFerence) is a reference to a value of some type T (e.g. Tensor) on a remote worker. This handle keeps the referenced remote value alive on the owner, but there is no implication that the value will be transferred to the local worker in the future. RRefs can be used in multi-machine training through holding references to `nn.Modules` that exist on other workers, and calling the appropriate functions to retrieve or modify their parameters during training.
```
Comment (mrshenli) [inline docs/source/rpc.rst]: Do we need to mark `RRef` and `T` as code, as this PR did for `ProcessGroup`?

File: docs/source/rpc.rst [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+------
+
+This library provides primitives allowing users to create and modify references to remote data and remotely execute functions and retrieve the result.
+
+.. automodule:: torch.distributed.rpc.api
+    :members:
+
+Distributed Autograd Framework
+====================================================
+This module provides an RPC-based distributed autograd framework that can be used for applications such as model parallel training. In short, applications may run RPC to execute code remotely in the forward pass, and autograd will automatically travel across RPC boundaries in the backward pass.
```
Comment (mrshenli) [inline docs/source/rpc.rst]: add a new line?

File: torch/csrc/distributed/rpc/init.cpp [hunk 9]
```diff
@@ -53,7 +53,9 @@ PyObject* rpc_init(PyObject* /* unused */) {
                           py::call_guard<py::gil_scoped_release>());
 
   auto pyRRef =
-      shared_ptr_class_<PyRRef>(module, "RRef")
+      shared_ptr_class_<PyRRef>(module, "RRef", R"(
+          A class encapsulating a reference to a value of some type on a remote worker. This handle will keep the referenced remote value alive on the worker.
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/init.cpp]: Shall we break this lone line into shorter ones?

**Review by mrshenli — APPROVED** [review]
The content LGTM! Please also get a stamp from @jlin27 on whether we should hide this for now and whether the structure is OK. Thanks!

File: docs/source/rpc.rst [hunk 10]
```diff
@@ -0,0 +1,42 @@
+.. role:: hidden
```
Comment (pritamdamania87) [inline docs/source/rpc.rst]: It looks like these titles show up in the sidebar:

```
Distributed RPC Framework
RPC and RRef Framework
Distributed Autograd Framework
```

Shouldn't we only have `Distributed RPC Framework` in the sidebar?

File: docs/source/rpc.rst [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+The distributed RPC framework provides mechanisms for multi-machine model training through a set of primitives to allow for remote communication, and a higher-level API to automatically differentiate models split across several machines.
+
+RPC and RRef Framework
+====================================================
+
+Basics
+------
+
+Before using RPC and distributed autograd primitives, initialization must take place. First, a backend over which RPCs can be sent over must be initialized. The default (and currently, only available) implementation is the `ProcessGroup` backend, and must be initialized with `torch.distributed.init_process_group` before using other function. See the `documentation for torch.distributed <https://pytorch.org/docs/stable/distributed.html>`_ for additional details. Next, the local RPC agent can be initialized, after which the process will be able to send and receive RPCs from all other connected processes.
```
Comment (pritamdamania87) [inline docs/source/rpc.rst]: Link to the docs for `torch.distributed.init_process_group` when you mention it. Also, shouldn't it be `PROCESS_GROUP` instead of `ProcessGroup`?

File: docs/source/rpc.rst [hunk 12]
```diff
@@ -0,0 +1,42 @@
+.. role:: hidden
+    :class: hidden-section
+
+Distributed RPC Framework
+=====================================================
+
+The distributed RPC framework provides mechanisms for multi-machine model training through a set of primitives to allow for remote communication, and a higher-level API to automatically differentiate models split across several machines.
```
Comment (pritamdamania87) [inline docs/source/rpc.rst]: Why do we have everything on a single long line? Can we split the lines in this file? Looks like most doc files split lines after 80 chars.

File: docs/source/rpc.rst (hunk 11 above)
Comment (pritamdamania87) [reply docs/source/rpc.rst]: We talk about initialization here, but the first doc below is for `RRef` instead of `init_model_parallel`?

Comment (pritamdamania87) [reply docs/source/rpc.rst]: > Next, the local RPC agent can be initialized, after which the process will be able to send and receive RPCs from all other connected processes.

Replace this with:

Next, to initialize the RPC framework we need to use `init_mode_parallel` which would initialize the RPC framework, RRef framework and distributed autograd.

File: docs/source/rpc.rst (hunk 7 above)
Comment (pritamdamania87) [inline docs/source/rpc.rst]: s/through holding/by holding/g, add a link to nn.Module

File: docs/source/rpc.rst [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+Basics
+------
+
+Before using RPC and distributed autograd primitives, initialization must take place. First, a backend over which RPCs can be sent over must be initialized. The default (and currently, only available) implementation is the `ProcessGroup` backend, and must be initialized with `torch.distributed.init_process_group` before using other function. See the `documentation for torch.distributed <https://pytorch.org/docs/stable/distributed.html>`_ for additional details. Next, the local RPC agent can be initialized, after which the process will be able to send and receive RPCs from all other connected processes.
+
+.. automodule:: torch.distributed.rpc
+    :members:
+
+RRef
```
Comment (pritamdamania87) [inline docs/source/rpc.rst]: Looks like the RRef class appears twice? 
Screen Shot 2019-11-11 at 12 41 58 PM

File: docs/source/rpc.rst (hunk 8 above)
Comment (pritamdamania87) [inline docs/source/rpc.rst]: Rewrite this as following:

In short, applications may send and receive gradient recording tensors over RPC. In the forward pass, we record when gradient recording tensors are sent over RPC and during the backward pass we use this information to perform a distributed backward pass using RPC. For more details see the design doc here (link to doc in https://github.com/pytorch/pytorch/pull/29175 once landed)

File: docs/source/rpc.rst (hunk 10 above)
Comment (rohan-varma) [reply docs/source/rpc.rst]: This is now fixed.

File: docs/source/rpc.rst (hunk 13 above)
Comment (rohan-varma) [reply docs/source/rpc.rst]: This is fixed, I had to manually specify `rpc.init_model_parallel` as an autofunction to get the desired order.

File: docs/source/rpc.rst (hunk 10 above)
Comment (rohan-varma) [reply docs/source/rpc.rst]: This is fixed in the latest update.

**Comment by rohan-varma** [comment]
@pritamdamania87 All the comments are addressed, could you take another look? Thanks!

File: docs/source/rpc.rst [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+Distributed Autograd Framework
+------------------------------
+
+This module provides an RPC-based distributed autograd framework that can be
+used for applications such as model parallel training. In short, applications
+may send and receive gradient recording tensors over RPC. In the forward pass,
+we record when gradient recording tensors are sent over RPC and during the
+backward pass we use this information to perform a distributed backward pass
+using RPC. For more details see the design doc
+`here <https://github.com/pytorch/pytorch/pull/29175>`_.
```
Comment (rohan-varma) [inline docs/source/rpc.rst]: Once the linked PR is landed this can be updated.

File: docs/source/rpc.rst (hunk 11 above)
Comment (pritamdamania87) [reply docs/source/rpc.rst]: nit: We still have `ProcessGroup` instead of `PROCESS_GROUP` in "The default (and currently, only available) implementation is the `ProcessGroup` backend"

File: docs/source/rpc.rst (hunk 13 above)
Comment (pritamdamania87) [reply docs/source/rpc.rst]: Looks like the `RRef` class doesn't show up at all now.
Screen Shot 2019-11-11 at 7 22 09 PM

**Review by pritamdamania87 — APPROVED** [review]
Looks good to me, although please check why the `RRef` class doesn't show up and see if we can fix it before landing.

**Comment by rohan-varma** [comment]
@jlin27 Could you please take a look at the structure and whether it should be hidden before I land this? Thanks!

File: docs/source/rpc.rst [hunk 15]
```diff
@@ -0,0 +1,65 @@
+.. role:: hidden
+    :class: hidden-section
+
+Distributed RPC Framework
```
Comment (mrshenli) [inline docs/source/rpc.rst]: Let's mark this as experimental for now. Check out [this page](https://pytorch.org/docs/stable/named_tensor.html) as an example.

Comment (rohan-varma) [reply docs/source/rpc.rst]: Will do this in follow up PR

**Comment by rohan-varma** [comment]
Landing since all of the failures are pre existing and not related to this PR.
